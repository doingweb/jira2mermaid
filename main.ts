import { readCsv } from "./lib/csv.ts";
import { generateMermaid } from "./lib/mermaid.ts";

if (import.meta.main) {
  await main();
}

async function main() {
  if (Deno.args.length !== 1) {
    console.error(
      "Usage: jira2mermaid <csv_file_path>",
      "\n\nTurns a Jira issue search CSV export into a Mermaid diagram showing blocking dependencies."
    );
    Deno.exit(1); // Exit with error code
  }

  console.log(await jira2mermaid(Deno.args[0]));
}

export async function jira2mermaid(csvFilePath: string) {
  const tickets = await readCsv(csvFilePath);
  return generateMermaid(tickets);
}
