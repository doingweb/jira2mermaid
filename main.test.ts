import { join } from "jsr:@std/path";
import { it } from "jsr:@std/testing/bdd";
import { expect } from "jsr:@std/expect";
import { assertSnapshot } from "jsr:@std/testing/snapshot";
import { jira2mermaid } from "./main.ts";

it("renders Mermaid for the basic example CSV file (basic-example.csv)", async (context) => {
  const mermaid = await jira2mermaid(
    join(Deno.cwd(), "__fixtures__", "basic-example.csv")
  );

  await assertSnapshot(context, mermaid);
});

it('sanitizes the "Summary" field (weird-summaries.csv)', async (context) => {
  const mermaid = await jira2mermaid(
    join(Deno.cwd(), "__fixtures__", "weird-summaries.csv")
  );

  await assertSnapshot(context, mermaid);
});

it("just renders the boilerplate if there are no tickets (no-tickets.csv)", async (context) => {
  const mermaid = await jira2mermaid(
    join(Deno.cwd(), "__fixtures__", "no-tickets.csv")
  );

  await assertSnapshot(context, mermaid);
});

it("throws an error when the CSV file is empty (empty.csv)", async () => {
  await expect(
    jira2mermaid(join(Deno.cwd(), "__fixtures__", "empty.csv"))
  ).rejects.toThrow("CSV file is empty");
});
