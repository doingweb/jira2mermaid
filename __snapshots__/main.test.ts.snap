export const snapshot = {};

snapshot[`renders Mermaid for the basic example CSV file (basic-example.csv) 1`] = `
'%%{init: {"flowchart": {"defaultRenderer": "elk"}} }%%
graph LR
  classDef toDo fill:#f5f0e3,color:#5c544b,stroke:#5c544b;
  classDef inProgress fill:#e8f0fe,color:#3b5ba3,stroke:#3b5ba3;
  classDef pr fill:#f8e7ff,color:#5a3b80,stroke:#5a3b80;
  classDef done fill:#e6f4ea,color:#344a42,stroke:#344a42;
  PROJ-100[["PROJ-100<br>Story 1"]]
  PROJ-101("PROJ-101<br>Task 1")
  PROJ-102("PROJ-102<br>Task 2")
  PROJ-103(["PROJ-103<br>Sub-task 1"])
  PROJ-104(["PROJ-104<br>Sub-task 2"])
  PROJ-102 --> PROJ-101
  PROJ-103 --> PROJ-102
  PROJ-104 --> PROJ-102
  PROJ-103 -.-> PROJ-100
  PROJ-104 -.-> PROJ-100
  class PROJ-100 inProgress
  class PROJ-101,PROJ-102 done
  class PROJ-103 pr
  class PROJ-104 toDo'
`;

snapshot[`sanitizes the "Summary" field (weird-summaries.csv) 1`] = `
'%%{init: {"flowchart": {"defaultRenderer": "elk"}} }%%
graph LR
  classDef toDo fill:#f5f0e3,color:#5c544b,stroke:#5c544b;
  classDef inProgress fill:#e8f0fe,color:#3b5ba3,stroke:#3b5ba3;
  classDef pr fill:#f8e7ff,color:#5a3b80,stroke:#5a3b80;
  classDef done fill:#e6f4ea,color:#344a42,stroke:#344a42;
  PROJ-100[["PROJ-100<br>This #quot;summary#quot; has quotes"]]
  PROJ-101[["PROJ-101<br>This summary is super long and will get truncated ..."]]
  PROJ-102[["PROJ-102<br>This summary has a comma, which is fine"]]
  PROJ-103[["PROJ-103<br>This one has extra spaces around it"]]
  class PROJ-100,PROJ-101,PROJ-102,PROJ-103 done'
`;

snapshot[`just renders the boilerplate if there are no tickets (no-tickets.csv) 1`] = `
'%%{init: {"flowchart": {"defaultRenderer": "elk"}} }%%
graph LR
  classDef toDo fill:#f5f0e3,color:#5c544b,stroke:#5c544b;
  classDef inProgress fill:#e8f0fe,color:#3b5ba3,stroke:#3b5ba3;
  classDef pr fill:#f8e7ff,color:#5a3b80,stroke:#5a3b80;
  classDef done fill:#e6f4ea,color:#344a42,stroke:#344a42;'
`;
